pkg/serviceprovider/gitlab/metadataprovider.go

```diff
@@ -0,0 +1,108 @@
+package gitlab
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/hashicorp/go-retryablehttp"
+	api "github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/logs"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/serviceprovider"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/spi-shared/tokenstorage"
+	"github.com/xanzy/go-gitlab"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type metadataProvider struct {
+	tokenStorage    tokenstorage.TokenStorage
+	httpClient      *http.Client
+	glClientBuilder gitlabClientBuilder
+	baseUrl         string
+}
+
+func (p metadataProvider) Fetch(ctx context.Context, token *api.SPIAccessToken) (*api.TokenMetadata, error) {
+	lg := log.FromContext(ctx, "tokenName", token.Name, "tokenNamespace", token.Namespace)
+
+	data, err := p.tokenStorage.Get(ctx, token)
+	if err != nil {
+		lg.Error(err, "failed to get the token metadata")
+		return nil, fmt.Errorf("failed to get the token metadata: %w", err)
+	}
+	if data == nil {
+		return nil, nil
+	}
+
+	state := &TokenState{}
+
+	glClient, err := p.glClientBuilder.createAuthenticatedGlClient(ctx, token, p.baseUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create authenticated GitHub client: %w", err)
+	}
+
+	//if err := (&AllAccessibleRepos{}).FetchAll(ctx, ghClient, data.AccessToken, state); err != nil {
+	//	return nil, err
+	//}
+
+	username, userId, scopes, err := p.fetchUserAndScopes(ctx, glClient)
+	if err != nil {
+		return nil, err
+	}
+
+	js, err := json.Marshal(state)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling the state: %w", err)
+	}
+
+	metadata := &api.TokenMetadata{}
+
+	metadata.UserId = userId
+	metadata.Username = username
+	metadata.Scopes = scopes
+	metadata.ServiceProviderState = js
+
+	return metadata, nil
+}
+
+// fetchUserAndScopes fetches the scopes and the details of the user associated with the context
+func (p metadataProvider) fetchUserAndScopes(ctx context.Context, gitlabClient *gitlab.Client) (userName string, userId string, scopes []string, err error) {
+	lg := log.FromContext(ctx)
+	usr, resp, err := gitlabClient.Users.CurrentUser(gitlab.WithContext(ctx))
+	if err != nil {
+		lg.Error(err, "error during fetching user metadata from GitLab")
+		err = fmt.Errorf("failed to fetch user info: %w", err)
+		return
+	}
+	if resp.StatusCode != 200 {
+		lg.Error(err, "error during fetching user metadata from GitLab", "status", resp.StatusCode)
+		return "", "", nil, nil
+	}
+
+	type tokenResponseBody struct {
+		scope []string
+	}
+	responseBody := tokenResponseBody{}
+
+	req, err := retryablehttp.NewRequestWithContext(ctx, "GET", p.baseUrl+"/oauth/token/info", nil)
```

Comment:
Hmm... so the scopes of the user are "global" and don't differ per-repository?

Reply:
If you mean something like the Viewer permission on GitHub, I haven't seen something like it on GitLab.

Reply:
So how are we going to figure out if a user has access to one repository as opposed to another?

Reply:
You are right. I will have to do more research on how this can be achieved with GitLab and implement it into TokenState. However, as this is supposed to be an MVP, I would like to merge this PR without it and afterward create a new PR with the implementation.